File: app/src/main/java/com/bsstokes/bsdiy/db/BsDiyDatabase.java
package com.bsstokes.bsdiy.db;

import android.support.annotation.NonNull;

import com.bsstokes.bsdiy.api.DiyApi;

import org.jetbrains.annotations.NotNull;

import java.util.List;

import rx.Observable;

public interface BsDiyDatabase {
    Observable<List<Skill>> getAllSkills();

    Observable<Skill> getSkill(long skillId);

    Observable<Skill> getSkillByUrl(@NonNull String skillUrl);

    void putSkill(Skill skill);

    void putSkills(@NotNull List<Skill> skills);

    Observable<List<Challenge>> getChallenges(long skillId);

    Observable<Challenge> getDbChallenge(long challengeId);

    Observable<DiyApi.Challenge> getChallenge(long challengeId);

    void putChallenge(@NotNull Challenge challenge);

    void putChallenge(DiyApi.Challenge challenge, long skillId, int position);

    void putChallenges(List<DiyApi.Challenge> challenges, long skillId);
}

File: app/src/main/java/com/bsstokes/bsdiy/db/BsDiyDatabase.kt
package com.bsstokes.bsdiy.db

import rx.Observable

interface BsDiyDatabase {
    fun getAllSkills(): Observable<List<Skill>>

    fun getSkill(skillId: Long): Observable<Skill?>

    fun getSkillByUrl(skillUrl: String): Observable<Skill?>

    fun putSkill(skill: Skill)

    fun putSkills(skills: List<Skill>)

    fun getChallenges(skillId: Long): Observable<List<Challenge>>

    fun getChallenge(challengeId: Long): Observable<Challenge?>

    fun putChallenge(challenge: Challenge)

    fun putChallenges(challenges: List<Challenge>)
}
